pkg/image/imagecontentsources/alternativesources.go

```diff
@@ -0,0 +1,173 @@
+	var imageSources []imagereference.DockerImageReference
+	for _, icsp := range a.ICSPList {
+		repoDigestMirrors := icsp.Spec.RepositoryDigestMirrors
+		var sourceMatches bool
+		for _, rdm := range repoDigestMirrors {
+			rdmRef, err := imagereference.Parse(rdm.Source)
+			if err != nil {
+				return nil, err
+			}
+			if imageRef.AsRepository() == rdmRef.AsRepository() {
+				klog.V(2).Infof("%v RepositoryDigestMirrors source matches given image", imageRef.AsRepository())
+				sourceMatches = true
+			}
+			for _, m := range rdm.Mirrors {
+				if sourceMatches {
+					klog.V(2).Infof("%v RepositoryDigestMirrors mirror added to potential ImageSourcePrefixes from ImageContentSourcePolicy", m)
+					mRef, err := imagereference.Parse(m)
+					if err != nil {
+						return nil, err
+					}
+					imageSources = append(imageSources, mRef)
+				}
+			}
+		}
+	}
+	uniqueMirrors := make([]imagereference.DockerImageReference, 0, len(imageSources))
+	uniqueMap := make(map[imagereference.DockerImageReference]bool)
+	for _, imageSourceMirror := range imageSources {
+		if _, ok := uniqueMap[imageSourceMirror]; !ok {
+			uniqueMap[imageSourceMirror] = true
+			uniqueMirrors = append(uniqueMirrors, imageSourceMirror)
+		}
+	}
+	// make sure at least 1 imagesource
+	// ie, make sure the image passed is included in image sources
+	// this is so the user-given image ref will be tried
+	if len(imageSources) == 0 {
+		imageSources = append(imageSources, imageRef.AsRepository())
+		return imageSources, nil
+	}
+	klog.V(2).Infof("Found sources: %v for image: %v", uniqueMirrors, imageRef)
+	return uniqueMirrors, nil
+}
+
+type SourceList struct {
+	Sources []string
+}
+
+// Next() satisfies registryclient.ImageSources interface
+// Will return next imageref string
+func (a *SourceList) Next() (string, error) {
+	if len(a.Sources) == 0 {
+		return "", nil
+	}
+	current := a.Sources
+	newList := append(current[:0], current[0+1:]...)
+	a.Sources = newList
+	return current[0], nil
+}
+
+// Resolve returns a possible image sources for a given image
+// Satisfies registrycontext.ImageResolver interface
+func (a *ICSPSources) Resolve(image string) (registryclient.ImageSources, error) {
+	sources, err := a.alternative(image)
+	if err != nil {
+		return nil, err
+	}
+	return &SourceList{Sources: sources}, nil
+}
+
+// alternative returns possible DockerImageReferences
+// from image sources gathered from ImageContentSourcePolicy objects and user-passed image.
+// Will lookup from cluster or from ImageContentSourcePolicy file passed from user.
+// Image reference of user-given image may be different from original in case of mirrored images.
+func (a *ICSPSources) alternative(image string) ([]string, error) {
+	imageRef, err := imagereference.Parse(image)
+	if err != nil {
+		return nil, err
+	}
+	if err := a.addImageSourcePoliciesFromFile(); err != nil {
+		return nil, err
+	}
+	if len(a.ICSPFile) == 0 {
```

Comment:
To make it nicer, you could return from `addImageSourcePoliciesFromFile` information how many sources your read or boolean flag indicating if you read anything, and error. Just a suggestion :wink:

Reply:
> A suggestion, I'd split this PR into 3 commits, for readability:
> 
> 1. that 2 new interfaces that Oleg proposed
> 2. reference implementation with ICSPFile and ICSPClient
> 3. bump
>    This will make the reviews nicer.

haven't split yet, but I've found I only need the single `ImageResolver` interface with the method `ResolveNext` I've pushed this

Reply:
I think it's fine as/is, since whenever an icsp file is passed we want to error if no icsp obj was decoded. We don't want to  return nil if file was successfully accessed but no icsp was found, that should be always an error.